src/Catalyst.Core.Modules.Keystore/KeyStoreService.cs

```diff
@@ -0,0 +1,839 @@
+            var decryptionKey = await GetPrivateKeyAsync(recipient.key.Name, cancel).ConfigureAwait(false);
+            return recipient.recipient.GetContent(decryptionKey);
+        }
+
+        /// <summary>
+        ///     Get the key ID for a recipient.
+        /// </summary>
+        /// <param name="ri">
+        ///     A recepient of the message.
+        /// </param>
+        /// <returns>
+        ///     The key ID of the recepient or <b>null</b> if the recepient info
+        ///     is not understood or does not contain an IPFS key id.
+        /// </returns>
+        /// <remarks>
+        ///     The key ID is either the Subject Key Identifier (preferred) or the
+        ///     issuer's distinguished name with the form "CN=&lt;kid>,OU=keystore,O=ipfs".
+        /// </remarks>
+        private MultiHash GetKeyId(RecipientInformation ri)
+        {
+            // Any errors are simply ignored.
+            try
+            {
+                // Subject Key Identifier is the key ID.
+                if (ri.RecipientID.SubjectKeyIdentifier is byte[] ski)
+                {
+                    return new MultiHash(ski);
+                }
+
+                // Issuer is CN=<kid>,OU=keystore,O=ipfs
+                var issuer = ri.RecipientID.Issuer;
+                if (issuer != null && issuer.GetValueList(X509Name.OU).Contains("keystore")
+                 && issuer.GetValueList(X509Name.O).Contains("ipfs"))
+                {
+                    var cn = issuer.GetValueList(X509Name.CN)[0] as string;
+                    return new MultiHash(cn);
+                }
+            }
+            catch (Exception e)
+            {
+                log.Warn("Failed reading CMS recipient info.", e);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        ///     Sets the passphrase for the key chain.
+        /// </summary>
+        /// <param name="passphrase"></param>
+        /// <param name="cancel">
+        ///     Is used to stop the task.  When cancelled, the <see cref="TaskCanceledException" /> is raised.
+        /// </param>
+        /// <returns>
+        ///     A task that represents the asynchronous operation.
+        /// </returns>
+        /// <exception cref="UnauthorizedAccessException">
+        ///     When the <paramref name="passphrase" /> is wrong.
+        /// </exception>
+        /// <remarks>
+        ///     The <paramref name="passphrase" /> is used to generate a DEK (derived encryption
+        ///     key).  The DEK is then used to encrypt the stored keys.
+        ///     <para>
+        ///         Neither the <paramref name="passphrase" /> nor the DEK are stored.
+        ///     </para>
+        /// </remarks>
+        public async Task SetPassphraseAsync(SecureString passphrase,
+            CancellationToken cancel = default)
+        {
+            // TODO: Verify DEK options.
+            // TODO: get digest based on Options.Hash
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)